src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -491,13 +492,32 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
             CancellationTokenSource cts;
             bool disposeCts;
             bool hasTimeout = _timeout != s_infiniteTimeout;
+            TimeoutState timeoutState = null; // Lazy initialization to avoid unnecessary alocation when timeout is not set
+            Timer timer = null;
             if (hasTimeout || cancellationToken.CanBeCanceled)
             {
                 disposeCts = true;
                 cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
                 if (hasTimeout)
                 {
-                    cts.CancelAfter(_timeout);
+                    long timeoutMilliseconds = (long)_timeout.TotalMilliseconds;
+                    if (timeoutMilliseconds < -1 || timeoutMilliseconds > int.MaxValue)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(_timeout));
+                    }
+
+                    timeoutState = new TimeoutState(cts);
+                    timer = new Timer(s =>
```

Comment:
What's keeping this rooted?  You're probably going to need to add this Timer as a field to TimeoutState, such that the Timer ends up rooting itself.  Otherwise, it looks like if garbage collection kicks in during the operation, the timer could get collected and never fire.  We should ensure we have a test for that.

Reply:
I thought about this problem and made the timer rooted by passing it to both of FinishSendAsync* methods where it gets passed down to HandleFinishSendAsyncCleanup. HandleFinishSendAsyncCleanup signature has been changed, so there is no way currently to call it without the time.
Additionally, I considered rooting it from the callback, but not sure if it's reliable since GC can obviously detect cycles.
Do you think the latter option is better?

Reply:
[I added a new test](https://github.com/dotnet/runtime/blob/29aaa4aaffd7176474325e2032aad723ae0ba02c/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs#L621) to make sure Timer is reliably rooted.